CMakeLists.txt

```diff
@@ -624,30 +620,43 @@ if(AVIF_BUILD_APPS OR (AVIF_BUILD_TESTS AND (AVIF_ENABLE_FUZZTEST OR AVIF_ENABLE
     endif()
 
     add_library(
-        avif_apps STATIC apps/shared/avifexif.c apps/shared/avifjpeg.c apps/shared/avifpng.c apps/shared/avifutil.c
-                         apps/shared/iccmaker.c apps/shared/y4m.c third_party/iccjpeg/iccjpeg.c
+        avif_apps_obj OBJECT $<TARGET_OBJECTS:avif_obj> apps/shared/avifexif.c apps/shared/avifjpeg.c apps/shared/avifpng.c
+                             apps/shared/avifutil.c apps/shared/iccmaker.c apps/shared/y4m.c third_party/iccjpeg/iccjpeg.c
     )
-    target_link_libraries(avif_apps avif PNG::PNG ZLIB::ZLIB JPEG::JPEG)
+    target_link_libraries(avif_apps_obj avif_obj PNG::PNG ZLIB::ZLIB JPEG::JPEG)
     if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
-        target_link_libraries(avif_apps m)
+        target_link_libraries(avif_apps_obj m)
     endif()
     # In GitHub CI's macos-latest os image, /usr/local/include has not only the headers of libpng
     # and libjpeg but also the headers of an older version of libavif. Put the avif include
     # directory before ${PNG_PNG_INCLUDE_DIR} ${JPEG_INCLUDE_DIR} to prevent picking up old libavif
     # headers from /usr/local/include.
-    target_include_directories(avif_apps PRIVATE $<TARGET_PROPERTY:avif,INTERFACE_INCLUDE_DIRECTORIES> third_party/iccjpeg)
-    target_include_directories(avif_apps SYSTEM PRIVATE ${PNG_PNG_INCLUDE_DIR} ${JPEG_INCLUDE_DIR})
-    target_include_directories(avif_apps INTERFACE apps/shared)
+    target_include_directories(avif_apps_obj PRIVATE third_party/iccjpeg)
+    target_include_directories(avif_apps_obj SYSTEM PRIVATE ${PNG_PNG_INCLUDE_DIR} ${JPEG_INCLUDE_DIR})
 
     if(AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP)
         if(TARGET LibXml2::LibXml2)
             set(AVIF_ENABLE_EXPERIMENTAL_JPEG_GAIN_MAP_CONVERSION TRUE)
             add_compile_definitions(AVIF_ENABLE_EXPERIMENTAL_JPEG_GAIN_MAP_CONVERSION)
-            target_link_libraries(avif_apps LibXml2::LibXml2)
+            target_link_libraries(avif_apps_obj LibXml2::LibXml2)
         else()
             message(STATUS "libavif: libxml2 not found; avifenc will ignore any gain map in jpeg files")
         endif()
     endif()
+
+    # Main avif_apps library.
+    add_library(avif_apps STATIC)
+    target_link_libraries(avif_apps PUBLIC avif PRIVATE avif_apps_obj)
+    target_include_directories(avif_apps INTERFACE apps/shared)
+
+    # avif_apps_internal is to use when linking to avif_internal.
+    if(BUILD_SHARED_LIBS)
+        add_library(avif_apps_internal STATIC)
+        target_link_libraries(avif_apps_internal PUBLIC avif_internal PRIVATE avif_apps_obj)
```

Comment:
I tested this on Windows. I found that this causes no problems because the static library avif_internal.lib can also satisfy the references to the `__imp_avif*` symbols. So I think we can ignore this issue.

I am a little confused now -- do we still need both the avif_apps and avif_apps_internal libraries? Or can we get by with just the avif_apps library?

Reply:
Indeed, I fixed it and `AVIF_DLL` is not defined for avif_apps anymore.